File: sort/count_sort.go
package sort

// count sort - asc
// input: 0 ~ MAX_INTEGER
func CountSort(inArr []uint) []uint {
	l := len(inArr)
	if l <= 1 {
		return inArr
	}
	maxValue, minValue := inArr[0], inArr[0]
	for i := 1; i < l; i++ {
		if maxValue < inArr[i] {
			maxValue = inArr[i]
		}
		if minValue > inArr[i] {
			minValue = inArr[i]
		}
	}
	tmpArr := make([]uint, maxValue+1)
	for i := 0; i < l; i++ {
		tmpArr[inArr[i]]++
	}
	result, idx := make([]uint, l), 0
	for i := minValue; i < uint(len(tmpArr)); i++ {
		count := tmpArr[i]
		for count > 0 {
			result[idx] = i
			count--
			idx++
		}
	}
	return result
}

File: sort/count_sort.cpp
// +build ignore
// from golang

#include "count_sort.h"

// count sort - asc
// input: 0 ~ MAX_INTEGER
std::vector<unsigned int> countSort(std::vector<unsigned int> inArr) {
    int size = inArr.size();
    if (size <= 1) return inArr;
    unsigned int maxValue=inArr[0], minValue = inArr[0];
    for(int i=1; i<size; ++i) {
        if(maxValue < inArr[i]) maxValue = inArr[i];
        if(minValue > inArr[i]) minValue = inArr[i];
    }
    std::vector<unsigned int> tmpArr(maxValue+1);
    for(int i=0; i<size; ++i) tmpArr[inArr[i]] += 1;
    std::vector<unsigned int> result;
    for(int i=minValue; i<int(tmpArr.size()); ++i) {
        int count = tmpArr[i];
        while(count > 0) {
            result.push_back(i);
            --count;
        }
    }
    return result;
}